README.md
# Scriptable Widgets

## Daily-Quote

> Shows you a motivating quote every new day.

![Screenshot](./Screenshots/Daily-Quote.jpg)

- Customizable settings:
  - `FONT`
  - `COLORS`
  - `API_TAGS` is a list of prefered tags for the quotes.
  - `SAVE_FILE` is the file to save the current quote in the *Scriptable* iCloud directory.

## Next-Spaceflight

> Shows you the dates of upcoming spaceflights.

![Screenshot](./Screenshots/Next-Spaceflight.jpg)

- Customizable settings:
  - `FONT`
  - `COLORS`
  - `API_PROVIDER` is a list of prefered companies of the launches.
  - `SAVE_FILE` is the file to save the current spaceflights in the *Scriptable* iCloud directory.

## Love-Counter

> Shows you the days since you have been with your partner.
> (actually very cheesy ;-)

![Screenshot](./Screenshots/Love-Counter.jpg)

- Customizable settings:
  - `FONT`
  - `COLORS`
  - `ANNIVERSARY` is the date you came together.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [v1.0] - 2023-09-23

- Initialized repository
- Added .gitignore
- Added README.md
- Added LICENSE.md
- Added CHANGELOG.md
- Added widgets: Daily-Quote, Next-Spaceflight, Love-Counter


Daily-Quote.js
// Variables used by Scriptable.
// These must be at the very top of the file. Do not edit.
// icon-color: deep-green; icon-glyph: book-open;
const API_URL = "https://api.quotable.io/random";
const API_TAGS = [];  // List of tags: https://api.quotable.io/tags

const FONT_NAME = "AmericanTypewriter";
const FONT_SIZE = 20;

const COLORS = {
  bg0: "#333333",  // Top
  bg1: "#4D4D44D",  // Bottom
  text0: "#FFFFFF",  // Content
  text1: "#FFFFFF"  // Author
};

const SAVE_FILE = "daily-quote.json";

if (config.runsInApp) {
  const data = await requestQuote();
  const widget = createWidget(data);
  console.log("> Running in app <")
  widget.presentMedium();
} else {
  const data = await fetchData();
  const widget = createWidget(data);
  Script.setWidget(widget);
}

console.log("Done!");
Script.complete();

function createWidget(data) {
  console.log("Creating widget...");

  // Base widget
  const widget = new ListWidget();
  const bgColor = new LinearGradient();
  bgColor.colors = [new Color(COLORS.bg0), new Color(COLORS.bg1)];
  bgColor.locations = [0.0, 1.0];
  widget.backgroundGradient = bgColor;
  widget.setPadding(10, 15, 15, 10);

  // Main stack
  const stack = widget.addStack();
  stack.layoutVertically();
  stack.spacing = 1;
  stack.size = new Size(0, 0);

  if (data != null) {
    // Add content
    const contentLine = stack.addText(data.content);
    contentLine.textColor = new Color(COLORS.text0);
    contentLine.font = new Font(FONT_NAME, FONT_SIZE);
    stack.addSpacer(10);

    // Add author
    const authorLine = stack.addText(`~ ${data.author}`);
    authorLine.textColor = new Color(COLORS.text1);
    authorLine.font = new Font(FONT_NAME, FONT_SIZE);
  } else {
    // Add info
    const infoLine = stack.addText("Connection error! 😿");
    infoLine.textColor = new Color(COLORS.text0);
    infoLine.font = new Font(FONT_NAME, FONT_SIZE);
  }

  return widget;
}

async function fetchData() {
  console.log("Loading file...")
  let data = null;

  const manager = FileManager.iCloud();
  const baseDir = manager.documentsDirectory();
  const file = manager.joinPath(baseDir, SAVE_FILE);

  const updatedDate = manager.modificationDate(file);
  const currentDate =  new Date();
  const dayValue = 24 * 60 * 60 * 1000;

  // Check if file needs to be rewritten
  if (!manager.fileExists(file) || currentDate - updatedDate >= dayValue) {
    data = await requestQuote();
    manager.writeString(file, JSON.stringify(data));
  } else {
    // Check if existent file needs to be downloaded
    if (!manager.isFileDownloaded(file)) {
      await manager.downloadFileFromiCloud(file);
    }

    data = JSON.parse(manager.readString(file));
  }

  return data;
}

async function requestQuote() {
  console.log("Requesting quote...");

  // Add tags to URL
  let url = API_URL + "?tags=";
  API_TAGS.forEach((element, index) => {
    url += element;
    if (index < API_TAGS.length - 1) {
      url += ";";
    }
  });

  // Fetch quote
  let data;
  try {
    const request = new Request(url);
    data = await request.loadJSON();
  } catch (error) {
    console.log("Connection error!")
    data = null;
  }

  return data;
}


Love-Counter.js
// Variables used by Scriptable.
// These must be at the very top of the file. Do not edit.
// icon-color: pink; icon-glyph: heart;
const NAMES = ["I", "N"];
const ANNIVERSARY = "2021-05-10";  // "Year-Month-Day"

const FONT_NAME = "Courier-Bold";
const FONT_SIZE = 25;

const COLORS = {
  bg0: "#000080",  // Top
  bg1: "#008080",  // Bottom
  line0: "#E81224",  // Foreground
  line1: "#EF6950",  // Background
  text: "#FFFFFF"
};

const CONTEXT_WIDTH = 125;
const CONTEXT_HEIGHT = 5;

const data = await fetchData();
const widget = createWidget(data);

if (config.runsInApp) {
  console.log("> Running in app <")
  widget.presentSmall();
}

console.log("Done!");
Script.setWidget(widget);
Script.complete();

function createWidget(data) {
  console.log("Creating Widget...");

  // Base widget
  const widget = new ListWidget();
  const bgColor = new LinearGradient();
  bgColor.colors = [new Color(COLORS.bg0), new Color(COLORS.bg1)];
  bgColor.locations = [0.0, 1.0];
  widget.backgroundGradient = bgColor;
  widget.setPadding(10, 15, 15, 10);

  // Main stack
  const stack = widget.addStack();
  stack.layoutVertically();
  stack.spacing = 1;
  stack.size = new Size(0, 0);

  // Add names
  const nameStack = stack.addStack();
  nameStack.addSpacer();
  const nameLine = nameStack.addText(`${NAMES[0]} ❤️ ${NAMES[1]}`);
  nameLine.textColor = new Color(COLORS.text);
  nameLine.font = new Font(FONT_NAME, FONT_SIZE);
  nameStack.addSpacer();
  stack.addSpacer(15);

  // Add progress line
  const progressStack = stack.addStack();
  progressStack.addSpacer();
  const progress = getProgressImage(data.progress);
  progressStack.addImage(progress);
  progressStack.addSpacer();
  stack.addSpacer(15);

  // Add day counter
  const dayStack = stack.addStack();
  dayStack.addSpacer();
  const dayLine = dayStack.addText(`${data.days} days`);
  dayLine.textColor = new Color(COLORS.text);
  dayLine.font = new Font(FONT_NAME, FONT_SIZE);
  dayStack.addSpacer();

  return widget;
}

function getProgressImage(value) {
  const context = new DrawContext();
  context.size = new Size(CONTEXT_WIDTH, CONTEXT_HEIGHT);
  context.respectScreenScale=true;
  context.opaque=false;

  // Background line
  const path = new Path();
  context.setFillColor(new Color(COLORS.line1));
  path.addRoundedRect(new Rect(0, 0, CONTEXT_WIDTH, CONTEXT_HEIGHT), 3, 2);
  context.addPath(path);
  context.fillPath();

  // Foreground line
  const path2 = new Path();
  context.setFillColor(new Color(COLORS.line0));
  path2.addRoundedRect(new Rect(0, 0, CONTEXT_WIDTH / 100 * value, CONTEXT_HEIGHT), 3, 2);
  context.addPath(path2);
  context.fillPath();

  return context.getImage();
}

async function fetchData() {
  const progress = new Date().getHours() * 100 / 24;
  const timeDifference = new Date().getTime() - new Date(ANNIVERSARY);
  const dayDifference = timeDifference / (1000 * 60 * 60 * 24);

  return {progress, days: `${Math.round(dayDifference)}`};
}


Next-Spaceflight.js
// Variables used by Scriptable.
// These must be at the very top of the file. Do not edit.
// icon-color: deep-blue; icon-glyph: rocket;
const API_URL = "https://fdo.rocketlaunch.live/json/launches/next/5";
const API_PROVIDER = ["SpaceX"]  // List of provider: https://www.rocketlaunch.live/api

const FONT_NAME = "Menlo-Bold";
const FONT_SIZE = 20;

const COLORS = {
  bg0: "#800080",  // Top
  bg1: "#000080",  // Bottom
  text: "#FFFFFF"
};

const SAVE_FILE = "space-flights.json";

if (config.runsInApp) {
  const data = await requestLaunches();
  const widget = createWidget(data);
  console.log("> Running in app <")
  widget.presentMedium();
} else {
  const data = await fetchData();
  const widget = createWidget(data);
  Script.setWidget(widget);
}

console.log("Done!");
Script.complete();

function createWidget(data) {
  console.log("Creating Widget...");

  // Base widget
  const widget = new ListWidget();
  const bgColor = new LinearGradient();
  bgColor.colors = [new Color(COLORS.bg0), new Color(COLORS.bg1)];
  bgColor.locations = [0.0, 1.0];
  widget.backgroundGradient = bgColor;
  widget.setPadding(10, 15, 15, 10);

  // Main stack
  const stack = widget.addStack();
  stack.layoutVertically();
  stack.spacing = 1;
  stack.size = new Size(0, 0);

  if (data != null) {
    // Add launches
    let text = "";
    data.forEach((element, index) => {
      text += `🚀 ${element.missions[0].name} - (${element.date_str})`;
      if (index < data.length - 1) {
        text += "\n";
      }
    });

    const launchLine = stack.addText(text);
    launchLine.textColor = new Color(COLORS.text);
    launchLine.font = new Font(FONT_NAME, FONT_SIZE);
    launchLine.minimumScaleFactor = 0.5;
  } else {
    // Add info
    const infoLine = stack.addText("Connection error! 😿");
    infoLine.textColor = new Color(COLORS.text);
    infoLine.font = new Font(FONT_NAME, FONT_SIZE);
  }

  return widget;
}

async function fetchData() {
  console.log("Loading file...")
  let data = null;

  const manager = FileManager.iCloud();
  const baseDir = manager.documentsDirectory();
  const file = manager.joinPath(baseDir, SAVE_FILE);

  const updatedDate = manager.modificationDate(file);
  const currentDate =  new Date();
  const dayValue = 24 * 60 * 60 * 1000;

  // Check if file needs to be rewritten
  if (!manager.fileExists(file) || currentDate - updatedDate >= dayValue) {
    data = await requestLaunches();
    manager.writeString(file, JSON.stringify(data));
  } else {
    // Check if existent file needs to be downloaded
    if (!manager.isFileDownloaded(file)) {
      await manager.downloadFileFromiCloud(file);
    }

    data = JSON.parse(manager.readString(file));
  }

  return data;
}

async function requestLaunches() {
  console.log("Requesting launches...");

  // Fetch launches
  let dataAll;
  try {
    const request = new Request(API_URL);
    dataAll = await request.loadJSON();
  } catch (error) {
    console.log("Connection error!")
    dataAll = null;
    return dataAll;
  }

  // Filter launches
  let dataFiltered = [];
  dataAll.result.forEach(element => {
    if (API_PROVIDER.some(item => item == element.provider.name)) {
      dataFiltered.push(element);
    }
  });

  return dataFiltered;
}
